{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script contains implementation of Global (Needleman-Wunsch), Fitting and Local (Smith Waterman) alignments along with some test examples. Note: This was done as a part of course on Bioinformatics at UIUC. Parts of the code and test examples have been adapted from the course content."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_blosum62(path):\n",
    "    \"\"\"\n",
    "    Reads in the ncbi's BLOSUM62.txt file and loads the scoring matrix\n",
    "    into a dictionary.\n",
    "    \n",
    "    :param: path is the full path in the local filesystem at which the .txt file is located\n",
    "    :return: a dictionary of dictionaries which will hold the cost of various amino acid\n",
    "    substitutions as defined in BLOSUM62.\n",
    "    \"\"\"\n",
    "    delta = {}\n",
    "    with open(path, 'r') as f:\n",
    "        #skipping the information in the header\n",
    "        lines = f.readlines()[6:]\n",
    "        keys = lines[0].split()\n",
    "        \n",
    "        #gap is labelled as '*' in the file, change it to standard '-'\n",
    "        keys[-1] = '-'\n",
    "        \n",
    "        #reading the scoring information for each amino acid and storing it as a dictionary\n",
    "        for i, line in enumerate(lines[1:]):\n",
    "            delta[keys[i]] = {k : int(v) for (k,v) in zip(keys, line.split()[1:])}  \n",
    "    \n",
    "    return delta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Needleman-Wunsch-Global Aligment\n",
    "import numpy as np\n",
    "UP = (-1,0)\n",
    "LEFT = (0, -1)\n",
    "TOPLEFT = (-1, -1)\n",
    "ORIGIN = (0, 0)\n",
    "point_list = [ORIGIN,UP,LEFT,TOPLEFT]\n",
    "\n",
    "def traceback_global(v, w, pointers):\n",
    "    \"\"\"\n",
    "    performs the traceback step of Needleman-Wunsch to obtain one possible alignment maximizing alignment score.\n",
    "    There may be other possible alignments that give the same optimal edit distance.\n",
    "    \"\"\"\n",
    "    \n",
    "    i,j = len(v), len(w)\n",
    "    #aligned sequences\n",
    "    new_v = []\n",
    "    new_w = []\n",
    "    \n",
    "    while True:\n",
    "        di, dj = pointers[i][j]\n",
    "        \n",
    "        #if there is a gap in sequence v, f(i,j) = f(i,j-1) + gap penalty during forward pass\n",
    "        if (di,dj) == LEFT:\n",
    "            new_v.append('-')\n",
    "            new_w.append(w[j-1])\n",
    "            \n",
    "        #if there is a gap in sequence w, f(i,j) = f(i-1,j) + gap penalty during forward pass\n",
    "        elif (di,dj) == UP:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append('-')\n",
    "            \n",
    "        #if it is a match or mismatch case, f(i,j) = f(i-1,j-1) + 0 or mismatch score during forward pass\n",
    "        elif (di,dj) == TOPLEFT:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append(w[j-1])\n",
    "            \n",
    "        i, j = i + di, j + dj\n",
    "        if (i <= 0 and j <= 0):\n",
    "            break\n",
    "            \n",
    "    #print(new_v, new_w)\n",
    "    return ''.join(new_v[::-1])+'\\n'+''.join(new_w[::-1])\n",
    "    \n",
    "\n",
    "\n",
    "def global_align(v, w, delta):\n",
    "    \"\"\"\n",
    "    Returns the score of the maximum scoring alignment of the strings v and w, as well as the actual alignment as \n",
    "    computed by traceback_global. \n",
    "    \n",
    "    :param: v\n",
    "    :param: w\n",
    "    :param: delta\n",
    "    \"\"\"\n",
    "    \n",
    "    #edit distance for substring v(0...i) and w(0...j)\n",
    "    M = [[0 for j in range(len(w)+1)] for i in range(len(v)+1)]\n",
    "    \n",
    "    #pointers needed for traceback step to get the optimal alignment\n",
    "    pointers = [[ORIGIN for j in range(len(w)+1)] for i in range(len(v)+1)]\n",
    "    \n",
    "    score, alignment = None, None\n",
    "    \n",
    "    #intialization for boundary cases\n",
    "    M[0][:] = [-j for j in range(len(w)+1)]\n",
    "    pointers[0][:] = [LEFT for j in range(len(w)+1)]\n",
    "    pointers[0][0] = ORIGIN\n",
    "    \n",
    "    #forward pass to obtian the edit distances\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(len(w)+1):\n",
    "            temp = [-float(\"inf\"),-float(\"inf\"),-float(\"inf\"),-float(\"inf\")]\n",
    "            if(i==0 and j==0):\n",
    "                temp[0] = 0\n",
    "            if(i>0):\n",
    "                temp[1] = M[i-1][j] + delta[v[i-1]]['-']\n",
    "            if(j>0):\n",
    "                temp[2] = M[i][j-1] + delta['-'][w[j-1]]\n",
    "            if(i>0 and j>0):\n",
    "                temp[3] = M[i-1][j-1] + delta[v[i-1]][w[j-1]]\n",
    "                    \n",
    "            pointers[i][j] = point_list[np.argmax(temp)]\n",
    "            M[i][j] = temp[np.argmax(temp)]\n",
    "    \n",
    "    #rightmost entry of bottom row gives the optimal edit distance/alignment score\n",
    "    score = M[len(v)][len(w)]\n",
    "    alignment = traceback_global(v,w, pointers)\n",
    "    return score, alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GIVEQCCTSICSLYQLENYCN\n",
      "GIVEQCCASVCSLYQLENYCN 115\n",
      "FVNQHLCGSHLVEALYLVCGERGFFYTPKA\n",
      "FVNQHLCGSHLVEALYLVCGERGFFYTPKA 167\n",
      "(5, 'LRRGEPVY----\\nLEKGDTLYILVG')\n",
      "(-2, 'ALYFFT\\nQCASVT')\n",
      "(-3, 'GTTTTATAGC-\\n-TGG-ATCGAT')\n"
     ]
    }
   ],
   "source": [
    "#Test cases for protein alignment\n",
    "a = 'GIVEQCCTSICSLYQLENYCN'\n",
    "b = 'FVNQHLCGSHLVEALYLVCGERGFFYTPKA'\n",
    "c = 'GIVEQCCASVCSLYQLENYCN'\n",
    "d = 'FVNQHLCGSHLVEALYLVCGERGFFYTPKA'\n",
    "\n",
    "blosum = read_blosum62('BLOSUM62.txt')\n",
    "\n",
    "scoreA,alignmentA = global_align(a,c,blosum)\n",
    "scoreB,alignmentB = global_align(b,d,blosum)\n",
    "print(alignmentA,scoreA)\n",
    "print(alignmentB, scoreB)\n",
    "\n",
    "print(global_align('LRRGEPVY', 'LEKGDTLYILVG', blosum))\n",
    "print(global_align('ALYFFT', 'QCASVT', blosum))\n",
    "\n",
    "# Test cases for DNA alignment\n",
    "# Scoring matrix (dictionary) for nucleotides\n",
    "keys = ['A', 'C', 'T', 'G', '-']\n",
    "delta = {}\n",
    "for i in range(len(keys)):\n",
    "    delta[keys[i]] = {k : v for (k,v) in zip(keys, [1 if keys[i] == keys[j]  else -1 for j in range(len(keys))])}\n",
    "\n",
    "print(global_align('GTTTTATAGC', 'TGGATCGAT', delta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fitting alignment\n",
    "import numpy as np\n",
    "UP = (-1,0)\n",
    "LEFT = (0, -1)\n",
    "TOPLEFT = (-1, -1)\n",
    "ORIGIN = (0, 0)\n",
    "point_list = [ORIGIN,UP,LEFT,TOPLEFT]\n",
    "\n",
    "def traceback_fitting(v, w, M, init_j, pointers):\n",
    "    \"\"\"\n",
    "    performs the traceback step of fitting alignment to obtain one possible alignment maximizing alignment score.\n",
    "    There may be other possible alignments that give the same optimal edit distance.\n",
    "    In addition to the pointers, we also need the position of the column from which to start the alignment.\n",
    "    \"\"\"\n",
    "    i, j = len(v), init_j\n",
    "    new_v = []\n",
    "    new_w = []\n",
    "    di,dj = (pointers[0][0])\n",
    "    \n",
    "    while True:\n",
    "        di, dj = pointers[i][j]\n",
    "        \n",
    "        #if there is a gap in sequence v, f(i,j) = f(i,j-1) + gap penalty during forward pass\n",
    "        if (di,dj) == LEFT:\n",
    "            new_v.append('-')\n",
    "            new_w.append(w[j-1])\n",
    "            \n",
    "        #if there is a gap in sequence w, f(i,j) = f(i-1,j) + gap penalty during forward pass\n",
    "        elif (di,dj) == UP:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append('-')\n",
    "            \n",
    "        #if it is a match or mismatch case, f(i,j) = f(i-1,j-1) + 0 or mismatch score during forward pass\n",
    "        elif (di,dj) == TOPLEFT:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append(w[j-1])\n",
    "        i, j = i + di, j + dj\n",
    "        if (i <= 0):\n",
    "            break\n",
    "    return ''.join(new_v[::-1]) + '\\n'+''.join(new_w[::-1])\n",
    "\n",
    "def fitting_align(short, reference, delta):\n",
    "    \"\"\"\n",
    "    Returns the score of the maximum scoring alignment of short and all \n",
    "    substrings of reference. \n",
    "    \n",
    "    :param: short the shorter of the two strings we are trying to align    \n",
    "    :param: reference the longer string among whose substrings we are doing global alignment\n",
    "    :param: delta the scoring function for the alphabet of the two strings\n",
    "    \n",
    "    :returns: a tuple (score, alignment)\n",
    "    \"\"\"\n",
    "    \n",
    "    #edit distance for substring v(0...i) and w(0...j)\n",
    "    M = [[0 for j in range(len(reference)+1)] for i in range(len(short)+1)]    \n",
    "    \n",
    "    #pointers needed for traceback step to get the optimal alignment\n",
    "    pointers = [[ORIGIN for j in range(len(reference)+1)] for i in range(len(short)+1)] \n",
    "    \n",
    "    score = None\n",
    "    init_j = 0\n",
    "    \n",
    "    #forward pass to obtian the alignment score\n",
    "    for i in range(1,len(short)+1):\n",
    "        for j in range(len(reference)+1):\n",
    "            temp = [-float(\"inf\"),-float(\"inf\"),-float(\"inf\"),-float(\"inf\")]\n",
    "            if(i==0 and j==0):\n",
    "                temp[0] = 0\n",
    "            if(i>0):\n",
    "                temp[1] = M[i-1][j] + delta[short[i-1]]['-']\n",
    "            if(j>0):\n",
    "                temp[2] = M[i][j-1] + delta['-'][reference[j-1]]\n",
    "            if(i>0 and j>0):\n",
    "                temp[3] = M[i-1][j-1] + delta[short[i-1]][reference[j-1]]\n",
    "            pointers[i][j] = point_list[np.argmax(temp)]     \n",
    "            M[i][j] = max(temp)\n",
    "            \n",
    "    #start of alignment is at the column with maximum score in the bottom column\n",
    "    score = max(M[len(short)][:])\n",
    "    max_idx = [i for i,j in enumerate(M[len(short)][:]) if j == score]\n",
    "    init_j = max_idx[-1]\n",
    "    \n",
    "    alignment = traceback_fitting(short,reference,M, init_j,pointers)\n",
    "    return score, alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5, 'AC-TGCT\\nACGTGCT')\n",
      "(24, 'PP--PVP\\nPPGKPVP')\n"
     ]
    }
   ],
   "source": [
    "# scoring matrix\n",
    "keys = ['A', 'C', 'T', 'G', '-']\n",
    "delta_fitting = {}\n",
    "for i in range(len(keys)):\n",
    "    delta_fitting[keys[i]] = {k : v for (k,v) in zip(keys, [1 if keys[i] == keys[j]  else -1 for j in range(len(keys))])}\n",
    "delta_fitting['-']['-'] = -1\n",
    "\n",
    "blosum = read_blosum62('BLOSUM62.txt')\n",
    "# Test cases\n",
    "\n",
    "print(fitting_align('ACTGCT', 'ACTGTCGTACGTGTACGTGCTATTACGATTCGGATGCATTGTGCATTTGGGCGATCTTATTCTTATC', delta))\n",
    "print(fitting_align('PPPVP','MHHHHHHSSPIDPPGKPVPLNITRHTVTLKWAKPEYTGGFKITSYIVEKRDLPNGRWLKANFSNILENEFTVSGLTEDAA\\\n",
    "YEFRVIAKNAAGAISPPSEPSDAITCRDDVEA', blosum))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Smith Waterman-Local alignment\n",
    "import numpy as np\n",
    "UP = (-1,0)\n",
    "LEFT = (0, -1)\n",
    "TOPLEFT = (-1, -1)\n",
    "ORIGIN = (0, 0)\n",
    "point_list = [ORIGIN,UP,LEFT,TOPLEFT]\n",
    "\n",
    "def traceback_local(v, w, M, init_i, init_j, pointers):\n",
    "    i,j = init_i, init_j\n",
    "    new_v = []\n",
    "    new_w = []\n",
    "    while True:\n",
    "        di, dj = pointers[i][j]\n",
    "        if (di,dj) == LEFT:\n",
    "            new_v.append('-')\n",
    "            new_w.append(w[j-1])\n",
    "        elif (di,dj) == UP:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append('-')\n",
    "        elif (di,dj) == TOPLEFT:\n",
    "            new_v.append(v[i-1])\n",
    "            new_w.append(w[j-1])\n",
    "        i, j = i + di, j + dj\n",
    "        if (M[i][j] == 0):\n",
    "            break\n",
    "    return ''.join(new_v[::-1]) + '\\n'+''.join(new_w[::-1])\n",
    "\n",
    "def local_align(v, w, delta):\n",
    "    \"\"\"\n",
    "    Returns the score of the maximum scoring alignment of all possible substrings of v and w. \n",
    "    \n",
    "    :param: v\n",
    "    :param: w\n",
    "    :param: delta\n",
    "    \"\"\"\n",
    "    M = [[0 for j in range(len(w)+1)] for i in range(len(v)+1)]        \n",
    "    pointers = [[ORIGIN for j in range(len(w)+1)] for i in range(len(v)+1)] \n",
    "    score = None\n",
    "    init_i, init_j = 0,0\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(len(w)+1):\n",
    "            temp = [0,-float(\"inf\"),-float(\"inf\"),-float(\"inf\")]\n",
    "            if(i>0):\n",
    "                temp[1] = M[i-1][j] + delta[v[i-1]]['-']\n",
    "            if(j>0):\n",
    "                #print(i,j)\n",
    "                temp[2] = M[i][j-1] + delta['-'][w[j-1]]\n",
    "            if(i>0 and j>0):\n",
    "                temp[3] = M[i-1][j-1] + delta[v[i-1]][w[j-1]]\n",
    "            pointers[i][j] = point_list[np.argmax(temp)]       \n",
    "            M[i][j] = max(temp)\n",
    "    score = np.max(M)\n",
    "    max_pos = np.where(M == score)\n",
    "    max_idx = np.argmax(max_pos[1])\n",
    "    init_i = (max_pos[0])[max_idx]\n",
    "    init_j = (max_pos[1])[max_idx]\n",
    "    alignment = traceback_local(v, w, M, init_i, init_j , pointers)\n",
    "    return score, alignment "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2355\n"
     ]
    }
   ],
   "source": [
    "LTK_MOUSE = \"MGCSHRLLLWLGAAGTILCSNSEFQTPFLTPSLLPVLVLNSQEQKVTPTP\\\n",
    "SKLEPASLPNPLGTRGPWVFNTCGASGRSGPTQTQCDGAYTGSSVMVTVG\\\n",
    "AAGPLKGVQLWRVPDTGQYLISAYGAAGGKGAQNHLSRAHGIFLSAVFFL\\\n",
    "RRGEPVYILVGQQGQDACPGGSPESQLVCLGESGEHATTYGTERIPGWRR\\\n",
    "WAGGGGGGGGATSIFRLRAGEPEPLLVAAGGGGRSYRRRPDRGRTQAVPE\\\n",
    "RLETRAAAPGSGGRGGAAGGGSGWTSRAHSPQAGRSPREGAEGGEGCAEA\\\n",
    "WAALRWAAAGGFGGGGGACAAGGGGGGYRGGDTSESDLLWADGEDGTSFV\\\n",
    "HPSGELYLQPLAVTEGHGEVEIRKHPNCSHCPFKDCQWQAELWTAECTCP\\\n",
    "EGTELAVDNVTCMDLPTTASPLILMGAVVAALALSLLMMCAVLILVNQKC\\\n",
    "QGLWGTRLPGPELELSKLRSSAIRTAPNPYYCQVGLSPAQPWPLPPGLTE\\\n",
    "VSPANVTLLRALGHGAFGEVYEGLVTGLPGDSSPLPVAIKTLPELCSHQD\\\n",
    "ELDFLMEALIISKFSHQNIVRCVGLSFRSAPRLILLELMSGGDMKSFLRH\\\n",
    "SRPHPGQLAPLTMQDLLQLAQDIAQGCHYLEENHFIHRDIAARNCLLSCS\\\n",
    "GASRVAKIGDFGMARDIYQASYYRKGGRTLLPVKWMPPEALLEGLFTSKT\\\n",
    "DSWSFGVLLWEIFSLGYMPYPGHTNQEVLDFIATGNRMDPPRNCPGPVYR\\\n",
    "IMTQCWQHQPELRPDFGSILERIQYCTQDPDVLNSPLPVEPGPILEEEEA\\\n",
    "SRLGNRSLEGLRSPKPLELSSQNLKSWGGGLLGSWLPSGLKTLKPRCLQP\\\n",
    "QNIWNPTYGSWTPRGPQGEDTGIEHCNGSSSSSIPGIQ\"\n",
    "\n",
    "ATK_MOUSE = \"MGAAGFLWLLPPLLLAAASYSGAATDQRAGSPASGPPLQPREPLSYSRLQ\\\n",
    "RKSLAVDFVVPSLFRVYARDLLLPQPRSPSEPEAGGLEARGSLALDCEPL\\\n",
    "LRLLGPLPGISWADGASSPSPEAGPTLSRVLKGGSVRKLRRAKQLVLELG\\\n",
    "EETILEGCIGPPEEVAAVGILQFNLSELFSWWILHGEGRLRIRLMPEKKA\\\n",
    "SEVGREGRLSSAIRASQPRLLFQIFGTGHSSMESPSETPSPPGTFMWNLT\\\n",
    "WTMKDSFPFLSHRSRYGLECSFDFPCELEYSPPLHNHGNQSWSWRHVPSE\\\n",
    "EASRMNLLDGPEAEHSQEMPRGSFLLLNTSADSKHTILSPWMRSSSDHCT\\\n",
    "LAVSVHRHLQPSGRYVAQLLPHNEAGREILLVPTPGKHGWTVLQGRVGRP\\\n",
    "ANPFRVALEYISSGNRSLSAVDFFALKNCSEGTSPGSKMALQSSFTCWNG\\\n",
    "TVLQLGQACDFHQDCAQGEDEGQLCSKLPAGFYCNFENGFCGWTQSPLSP\\\n",
    "HMPRWQVRTLRDAHSQGHQGRALLLSTTDILASEGATVTSATFPAPMKNS\\\n",
    "PCELRMSWLIRGVLRGNVSLVLVENKTGKEQSRTVWHVATDEGLSLWQHT\\\n",
    "VLSLLDVTDRFWLQIVTWWGPGSRATVGFDNISISLDCYLTISGEEKMSL\\\n",
    "NSVPKSRNLFEKNPNKESKSWANISGPTPIFDPTVHWLFTTCGASGPHGP\\\n",
    "TQAQCNNAYQNSNLSVVVGSEGPLKGVQIWKVPATDTYSISGYGAAGGKG\\\n",
    "GKNTMMRSHGVSVLGIFNLEKGDTLYILVGQQGEDACPRANQLIQKVCVG\\\n",
    "ENNVIEEEIRVNRSVHEWAGGGGGGGGATYVFKMKDGVPVPLIIAAGGGG\\\n",
    "RAYGAKTETFHPERLESNSSVLGLNGNSGAAGGGGGWNDNTSLLWAGKSL\\\n",
    "LEGAAGGHSCPQAMKKWGWETRGGFGGGGGGCSSGGGGGGYIGGNAASNN\\\n",
    "DPEMDGEDGVSFISPLGILYTPALKVMEGHGEVNIKHYLNCSHCEVDECH\\\n",
    "MDPESHKVICFCDHGTVLADDGVSCIVSPTPEPHLPLSLILSVVTSALVA\\\n",
    "ALVLAFSGIMIVYRRKHQELQAMQMELQSPEYKLSKLRTSTIMTDYNPNY\\\n",
    "CFAGKTSSISDLKEVPRKNITLIRGLGHGAFGEVYEGQVSGMPNDPSPLQ\\\n",
    "VAVKTLPEVCSEQDELDFLMEALIISKFNHQNIVRCIGVSLQALPRFILL\\\n",
    "ELMAGGDLKSFLRETRPRPNQPTSLAMLDLLHVARDIACGCQYLEENHFI\\\n",
    "HRDIAARNCLLTCPGAGRIAKIGDFGMARDIYRASYYRKGGCAMLPVKWM\\\n",
    "PPEAFMEGIFTSKTDTWSFGVLLWEIFSLGYMPYPSKSNQEVLEFVTSGG\\\n",
    "RMDPPKNCPGPVYRIMTQCWQHQPEDRPNFAIILERIEYCTQDPDVINTA\\\n",
    "LPIEYGPVVEEEEKVPMRPKDPEGMPPLLVSPQPAKHEEASAAPQPAALT\\\n",
    "APGPSVKKPPGAGAGAGAGAGAGPVPRGAADRGHVNMAFSQPNPPPELHK\\\n",
    "GPGSRNKPTSLWNPTYGSWFTEKPAKKTHPPPGAEPQARAGAAEGGWTGP\\\n",
    "GAGPRRAEAALLLEPSALSATMKEVPLFRLRHFPCGNVNYGYQQQGLPLE\\\n",
    "ATAAPGDTMLKSKNKVTQPGP\"\n",
    "\n",
    "score_local, alignment_local = local_align(LTK_MOUSE, ATK_MOUSE, blosum)\n",
    "print(score_local)\n",
    "#print(alignment_local)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6, 'GTGACGTCTATCAT\\nGTG-CATGTATC-T')\n",
      "(13, 'ACTGATTTCGATGC-TGTA-CGTGACGTACGTTTA\\nACT-ATT--G-TGCATGTATC-TGAC-TA-GCT-A')\n"
     ]
    }
   ],
   "source": [
    "keys = ['A', 'C', 'T', 'G', '-']\n",
    "delta_local = {}\n",
    "for i in range(len(keys)):\n",
    "    delta_local[keys[i]] = {k : v for (k,v) in zip(keys, [1 if keys[i] == keys[j]  else -1 for j in range(len(keys))])}\n",
    "delta_local['-']['-'] = -1\n",
    "print(local_align(\"TACGTGACGTCTATCAT\", \"TTGTGCATGTATCTGAC\", delta_local))\n",
    "print(local_align(\"ACTGATTTCGATGCTGTACGTGACGTACGTTTATTCTATCAT\", \"TCTGACTGTGACTATACTATTGTGCATGTATCTGACTAGCTAG\", delta_local))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
